### PR TITLE
Add checks for invalid iterator when _ITERATOR_DEBUG_LEVEL is non-zero

### LINKED ISSUE #3281: <vector>: iterator debugging misses a check before dereferencing
**Description**
When iterator debugging is enabled and a vector iterator is invalidated, for example by a call to `erase` at or before the referenced element, an attempt to dereference the iterator causes an access violation.

**Command-line test case**
```
C:\Temp>type repro.cpp
#include <vector>

int main() {
  std::vector<int> v({0,1,2,3});
  auto iter = v.begin();
  v.erase(v.begin()); // invalidate iter
  return *iter; // access violation, rather than an assertion
}
c:\temp>cl /EHsc /Zi /MDd repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31935 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.cpp
Microsoft (R) Incremental Linker Version 14.34.31935.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:repro.exe
/debug
repro.obj

c:\temp>repro

c:\temp>echo %ERRORLEVEL%
-1073741819

c:\temp>
```
The error code is that of `STATUS_ACCESS_VIOLATION` 0xC0000005 ("The instruction at 0x%p referenced memory at 0x%p. The memory could not be %s.").

**Expected behavior**
I was hoping to have an assert failure from the iterator dereference in `<vector>`.
What happens is an access violation.
The affected code is:

```
45     _NODISCARD _CONSTEXPR20 reference operator*() const noexcept {
46 #if _ITERATOR_DEBUG_LEVEL != 0
47         const auto _Mycont = static_cast<const _Myvec*>(this->_Getcont());
48         _STL_VERIFY(_Ptr, "can't dereference value-initialized vector iterator");
49         _STL_VERIFY(
50             _Mycont->_Myfirst <= _Ptr && _Ptr < _Mycont->_Mylast, "can't dereference out of range vector iterator");
51 #endif // _ITERATOR_DEBUG_LEVEL != 0
52 
53        return *_Ptr;
54     }
```
The AV occurs at line 49/50 as `_Mycont` is `null`; this is becasue the call to `erase` that invalidated the iterator changed the fields in the base class `_Iterator_base12` to reflect this, such that _Getcont() now returns `null` by design:
In `<vector>`:
```
2216             } else { // orphan the iterator
2217                _Temp->_Myproxy = nullptr;
```

**STL version**
Microsoft Visual Studio Community 2022 (64-bit) - Current
Version 17.4.2

**Possible Proposed Resolution**
Add a new line between lines 48 and 49:
```
        _STL_VERIFY(_Mycont, "can't dereference invalidated vector iterator");
```
The check takes place _after_ the check on `_Ptr`, which detects ''uninitialised'' iterators, and before the problematic dereference of `_Mycont`.

A similar additional line could be added to other methods which perform similar checking, such as `operator->`.

I'm raising a P/R ...

### PR DESCRIPTION
**Pull request by rogerorr**
fixes #3281

### CODE REVIEW

**Comment by rogerorr** [comment]
Links with https://github.com/microsoft/STL/issues/3281

**Comment by fsb4000** [comment]
You can link properly, just change your message to
```
Fixes #3821
```

**Comment by rogerorr** [comment]
> > I don't believe I have the permissions required to link these "properly"
> 
> You can link properly, just change your message to
> 
> ```
> Fixes #3821
> ```
> 
> Github docs: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

Hmm, I've tried changing the message to both Fixes #3821 and fixes #3821 but still don't see a linking item in the right hand pane. What am I doing wrong?

**Comment by fsb4000** [comment]
Really strange. Look at my message: https://github.com/microsoft/STL/pull/3276#issue-1486983628
I don't know why
Fixes #3281
doesn't work for you...

**Comment by fsb4000** [comment]
Oh, my bad.
I wrote a wrong number for you.
you need write 3281, not 3821.
Sorry.

**Comment by rogerorr** [comment]
> Oh, my bad. I wrote a wrong number for you. you need write 3281, not 3821. Sorry.

Ach, and I didn't notice either. Now changed and working properly. Thank you!

File: stl/inc/vector [hunk 1]
```diff
@@ -46,6 +46,7 @@ public:
 #if _ITERATOR_DEBUG_LEVEL != 0
         const auto _Mycont = static_cast<const _Myvec*>(this->_Getcont());
         _STL_VERIFY(_Ptr, "can't dereference value-initialized vector iterator");
+        _STL_VERIFY(_Mycont, "can't dereference invalidated vector iterator");
```
Comment (strega-nil-ms) [inline stl/inc/vector:49]: Doesn't this subsume the previous line? Can we just check for `_STL_VERIFY(_Mycont` and not check for `_Ptr` as well?

Comment (rogerorr) [reply stl/inc/vector:49]: Well, we _could_ but I think there is a useful distinction to retain between attempting to dereference a _value initialized_ vector iterator and attempting to dereference an _invalided_ iterator.

Comment (rogerorr) [reply stl/inc/vector:49]: Ah, and additionally I believe `_Mycont` is only cleared at `_ITERATOR_DEBUG_LEVEL` of 2, so the check would be lost at level 1.

**Review by strega-nil-ms — APPROVED** [review]
I don't know enough about iterator debugging checks to say, but this does seem extremely reasonable. I would like someone who's been on the project longer than me to take a look, though.

**Review by CaseyCarter — APPROVED** [review]
In the past we've simply allowed these cases to crash by dereferencing a null `_Mycont`, but reporting specific error messages is probably much more digestible for non-expert users. Thanks Roger!

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
As @CaseyCarter noted, we previously removed a lot of our "we're about to dereference null" debugging checks because null dereferences are fairly easy to understand. I agree that it makes sense to add null checks here to help users, because these null dereferences are special - they would occur in our debugging machinery itself. I also checked, and in other containers we always test `_Mycont` before dereferencing it, so this is making `vector` consistent with them.

I noticed that there's one remaining case: https://github.com/microsoft/STL/blob/4483e87f6e5aa5c4598404fc01eecb7e547dd439/stl/inc/vector#L101-L115

This is used by `operator+=` et al. and it'll dereference null `_Mycont` if it's called on an invalidated `vector` iterator that ever pointed to an element (i.e. it had a non-null `_Ptr`) and it's being asked to seek non-zero:

```
D:\GitHub\STL\out\x64>type meow.cpp
```
```cpp
#include <cstdio>
#include <vector>
using namespace std;

int main() {
    vector<int>::iterator it;

    {
        vector<int> vec{11, 22, 33};
        it = vec.begin();
    }

    puts("Running it += 0;");
    it += 0;
    puts("... done.");

    puts("Running it += 1;");
    it += 1;
    puts("... done. (UH OH)");

    puts("Running it += -1;");
    it += -1;
    puts("... done. (ALSO UH OH)");
}
```
```
D:\GitHub\STL\out\x64>cl /EHsc /nologo /W4 /MTd meow.cpp && meow
meow.cpp
Running it += 0;
... done.
Running it += 1;
[***NULL DEREF***]
```

(Seeking an invalidated iterator by zero is technically bogus, but it's debatable whether we should complain here. Seeking by nonzero is super bogus.)

Ordinarily I would simply push changes to handle this case, but that might cause our Contributor License Agreement bot to ask you to agree to our CLA (it appears to have a de minimis threshold for tiny changes like a 5-line PR, but it'll ask for larger PRs). While our CLA is totally cool, on rare occasions it makes someone nervous, so I prefer to avoid pushing small PRs from new contributors over the limit if we can avoid it.

As this PR is a strict improvement, I think we can merge it as-is, and I've recorded a todo to update `_Verify_offset` later.

**Comment by rogerorr** [comment]
> Ordinarily I would simply push changes to handle this case, but that might cause our Contributor License Agreement bot to ask you to agree to our CLA (it appears to have a de minimis threshold for tiny changes like a 5-line PR, but it'll ask for larger PRs). While our CLA is totally cool, on rare occasions it makes someone nervous, so I prefer to avoid pushing small PRs from new contributors over the limit if we can avoid it.

I'd be fine with signing it; Iwas expecting it to trigger when I raised the P/R and wondered whether it was suppressed by the prior NDA I'd signed with MS
 
> As this PR is a strict improvement, I think we can merge it as-is, and I've recorded a todo to update `_Verify_offset` later.

Agree. Dereferencing is also I suspect more common than invoking operator+=

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for improving these runtime diagnostics, and congratulations on your first microsoft/STL commit! :tada: :rocket: :smile_cat:

This will be available in VS 2022 17.6 Preview 1.

**Comment by pbdot** [comment]
I don't have a minimal reproduction case yet, though this trips when using Clang in debug builds, where MSVC does not

**Pull request merged**